### PR TITLE
refactor: improve policy for exports field

### PR DESCRIPTION
**Pull request by tylerbutler**
Improves the policy checker and resolver that validates the exports field in package.json. Improvements include:

- Verifies that all non-private packages have an exports field.
- Exports field must have a `types` condition.
- Exports field must have a `require` and `import` condition unless package is ESM or CJS-only.
- Packages with `exports` should not have the `module` field set.
- Verifies that the `types` and `main` fields match the corresponding exports fields. The resolver unifies the fields with the exports values.
- Prefers the "simple" form of exports for ESM and CJS-only packages. That is, if the package is only a single format, the extra level of nesting under "default" is removed.
- ESM-only packages should have type: module.
- Any non-default exports (e.g. ./alpha, ./internal) are ignored and should be preserved by the resolver.

I pre-applied the changes from the new resolver to the client release group in #20523. The main problem was that the types fields didn't match.

I also manually updated some build-tools packages that are CJS-only to set type: commonjs.

### CODE REVIEW

File: build-tools/packages/version-tools/package.json [hunk 1]
```diff
@@ -20,7 +21,7 @@
 		}
 	},
 	"main": "lib/index.js",
-	"types": "lib/index.d.ts",
+	"types": "./lib/index.d.ts",
```
Comment (alexvy86) [inline build-tools/packages/version-tools/package.json:24]: I think I'd go without the leading dot but don't mind too much to be honest. However, whichever way, can we make it match the way "main" is specified?

Also worth noting that build-cli also acquired `type: commonjs` but didn't have its `types` line changed...?

**Review by alexvy86 — COMMENTED** [review]
Haven't looked at the logic yet, leaving a small comment for now.

File: build-tools/packages/version-tools/package.json (hunk 1 above)
Comment (tylerbutler) [reply build-tools/packages/version-tools/package.json:24]: > Also worth noting that build-cli also acquired `type: commonjs` but didn't have its `types` line changed...?

These were manually added. I'll update the PR description.

Comment (tylerbutler) [reply build-tools/packages/version-tools/package.json:22]: > I think I'd go without the leading dot but don't mind too much to be honest. However, whichever way, can we make it match the way "main" is specified?

The latest version now normalizes the main and types fields to match exports.

**Comment by tylerbutler** [comment]
Marking this as a draft since it may need to be configurable.

**Pull request closed without merging**